Unable to copy layered worldstate on 22.10.2 - Snapshots=False

### Description
Besu user experiencing the following bug:

022-12-07 08:38:20.491-07:00 | vert.x-worker-thread-0 | INFO  | MainnetBlockValidator | Optional[Unable to copy Layered Worldstate for 0x3c2a249675dca59a4477c9a3ffdcc51e203617266f1a0f6434fa1a1dc5fec3f0]. Block 16130345 (0xa8b549ea57d83df60aec71c6fb40ab8391b0f640d7a2d2a013fce3a296800ad0), caused by org.hyperledger.besu.plugin.services.exception.StorageException: Unable to copy Layered Worldstate for 0x3c2a249675dca59a4477c9a3ffdcc51e203617266f1a0f6434fa1a1dc5fec3f0...

Directly after a restart and their node is stuck. Restart does not fix. Initially the user experienced the error below, then upon restart, the error above. 

2022-12-07 08:09:44.904-07:00 | vert.x-worker-thread-0 | WARN  | EngineNewPayload | Invalid new payload: number: 16130345, hash: 0xa8b549ea57d83df60aec71c6fb40ab8391b0f640d7a2d2a013fce3a296800ad0, parentHash: 0x3c2a249675dca59a4477c9a3ffdcc51e203617266f1a0f6434fa1a1dc5fec3f0, latestValidHash: 0x3c2a249675dca59a4477c9a3ffdcc51e203617266f1a0f6434fa1a1dc5fec3f0, status: INVALID, validationError: Unable to process block because parent world state 0x3b34cb5ab01b1b590a73ac002acbf60790360952ed1679bd6d9f9b608856b29a is not available

User does not have snapshots enabled. User is on 22.10.2. 

[Discord context](https://discord.com/channels/697535391594446898/888098626264502302/1049027346226757682)

### Acceptance Criteria
* Bonsai is resilient to this concurrency bug 

### Steps to Reproduce (Bug)
1. Run Besu
2. Halt or restart
3. Experience error (specific DB state)

**Expected behavior:** [What you expect to happen]
Continues normally. 

**Actual behavior:** [What actually happens]
Bonsai resumes as normal

**Frequency:** [What percentage of the time does it occur?]
Infrequent

### Versions (Add all that apply)
* Software version: 22.10.2

### Additional Information (Add any of the following or anything else that may be relevant)
* Besu setup info - snapshots false
* System info - memory, CPU

I started getting this issue and tried the 23.1.0 release candidate docker image to mitigate it and am getting these errors instead:

```
rocketpool_eth1  | 2023-01-01 02:26:03.390+00:00 | vert.x-eventloop-thread-17 | ERROR | ExecutionEngineJsonRpcMethod | failed to exec consensus method engine_newPayloadV1
rocketpool_eth1  | java.lang.RuntimeException: java.lang.RuntimeException: org.hyperledger.besu.plugin.services.exception.StorageException: Unable to copy Layered Worldstate for 0x334a634a192f87ae76c23c1b8a1f40619382d00825877d07ebef1947d777f637
rocketpool_eth1  | 	at org.hyperledger.besu.ethereum.MainnetBlockValidator.validateAndProcessBlock(MainnetBlockValidator.java:175)
rocketpool_eth1  | 	at org.hyperledger.besu.consensus.merge.blockcreation.MergeCoordinator.validateBlock(MergeCoordinator.java:441)
rocketpool_eth1  | 	at org.hyperledger.besu.consensus.merge.blockcreation.MergeCoordinator.rememberBlock(MergeCoordinator.java:455)
rocketpool_eth1  | 	at org.hyperledger.besu.consensus.merge.blockcreation.TransitionCoordinator.rememberBlock(TransitionCoordinator.java:142)
rocketpool_eth1  | 	at org.hyperledger.besu.ethereum.api.jsonrpc.internal.methods.engine.EngineNewPayload.syncResponse(EngineNewPayload.java:219)
rocketpool_eth1  | 	at org.hyperledger.besu.ethereum.api.jsonrpc.internal.methods.ExecutionEngineJsonRpcMethod.lambda$response$0(ExecutionEngineJsonRpcMethod.java:73)
rocketpool_eth1  | 	at io.vertx.core.impl.ContextImpl.lambda$null$0(ContextImpl.java:159)
rocketpool_eth1  | 	at io.vertx.core.impl.AbstractContext.dispatch(AbstractContext.java:100)
rocketpool_eth1  | 	at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$1(ContextImpl.java:157)
rocketpool_eth1  | 	at io.vertx.core.impl.TaskQueue.run(TaskQueue.java:76)
rocketpool_eth1  | 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
rocketpool_eth1  | 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
rocketpool_eth1  | 	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
rocketpool_eth1  | 	at java.base/java.lang.Thread.run(Thread.java:833)
rocketpool_eth1  | Caused by: java.lang.RuntimeException: org.hyperledger.besu.plugin.services.exception.StorageException: Unable to copy Layered Worldstate for 0x334a634a192f87ae76c23c1b8a1f40619382d00825877d07ebef1947d777f637
rocketpool_eth1  | 	at org.hyperledger.besu.ethereum.bonsai.BonsaiLayeredWorldState.copy(BonsaiLayeredWorldState.java:275)
rocketpool_eth1  | 	at org.hyperledger.besu.ethereum.MainnetBlockValidator.lambda$validateAndProcessBlock$0(MainnetBlockValidator.java:120)
rocketpool_eth1  | 	at java.base/java.util.Optional.map(Optional.java:260)
rocketpool_eth1  | 	at org.hyperledger.besu.ethereum.MainnetBlockValidator.validateAndProcessBlock(MainnetBlockValidator.java:117)
rocketpool_eth1  | 	... 13 more
rocketpool_eth1  | Caused by: org.hyperledger.besu.plugin.services.exception.StorageException: Unable to copy Layered Worldstate for 0x334a634a192f87ae76c23c1b8a1f40619382d00825877d07ebef1947d777f637
rocketpool_eth1  | 	at org.hyperledger.besu.ethereum.bonsai.BonsaiLayeredWorldState.lambda$copy$4(BonsaiLayeredWorldState.java:272)
rocketpool_eth1  | 	at java.base/java.util.Optional.orElseThrow(Optional.java:403)
rocketpool_eth1  | 	at org.hyperledger.besu.ethereum.bonsai.BonsaiLayeredWorldState.copy(BonsaiLayeredWorldState.java:269)
rocketpool_eth1  | 	... 16 more
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -7,6 +7,8 @@
 ### Additions and Improvements
 
 ### Bug Fixes
+- Mitigation fix for stale bonsai code storage leading to log rolling issues on contract recreates [#4906](https://github.com/hyperledger/besu/pull/4906)
+
 
 ## 23.1.0-RC1
 

--- a/ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/BonsaiWorldStateUpdater.java
+++ b/ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/BonsaiWorldStateUpdater.java
@@ -43,9 +43,12 @@ import org.apache.tuweni.bytes.Bytes;
 import org.apache.tuweni.bytes.Bytes32;
 import org.apache.tuweni.units.bigints.UInt256;
 import org.jetbrains.annotations.NotNull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 public class BonsaiWorldStateUpdater extends AbstractWorldUpdater<BonsaiWorldView, BonsaiAccount>
     implements BonsaiWorldView {
+  private static final Logger LOG = LoggerFactory.getLogger(BonsaiWorldStateUpdater.class);
 
   private final AccountConsumingMap<BonsaiValue<BonsaiAccount>> accountsToUpdate;
   private final Consumer<BonsaiValue<BonsaiAccount>> accountPreloader;
@@ -601,10 +604,8 @@ public class BonsaiWorldStateUpdater extends AbstractWorldUpdater<BonsaiWorldVie
       if ((expectedCode == null || expectedCode.isEmpty())
           && existingCode != null
           && !existingCode.isEmpty()) {
-        throw new IllegalStateException(
-            String.format("Expected to create code, but the code exists.  Address=%s", address));
-      }
-      if (!Objects.equals(expectedCode, existingCode)) {
+        LOG.warn("At Address={}, expected to create code, but code exists. Overwriting.", address);
+      } else if (!Objects.equals(expectedCode, existingCode)) {
         throw new IllegalStateException(
             String.format(
                 "Old value of code does not match expected value.  Address=%s ExpectedHash=%s ActualHash=%s",

